perf(snapshot): avoid recreate element `a` every time

H4ad opened the pull request:
Avoid recreating the same element every time, instead, we cache and we just update the element.

Before: 779k ops/s
After: 860k ops/s

Benchmark: https://jsbench.me/ktlqztuf95/1

On Chrome:
![image](https://github.com/rrweb-io/rrweb/assets/12551007/911a6055-81ae-461b-bf1c-7ada5e11dbec)

On Firefox:
![image](https://github.com/rrweb-io/rrweb/assets/12551007/5e5eda1e-c573-45cf-9c9b-402bd69e07e0)

JonasBa commented on packages/rrweb-snapshot/src/snapshot.ts:
```diff
@@ -194,23 +194,34 @@ function getAbsoluteSrcsetString(doc: Document, attributeValue: string) {
+    cachedDocument.set(doc, a);
+  }
+
+  // this is needed to work with SPA
```
Can you explain why this is needed for SPA? Folks reading this later might not necessarily have context or understanding why this is necessary :)

H4ad replied:
I will add a comment to link to this answer. 

Basically, the cached `a` link stores the URL when the element is created.
SPAs will not update the page but it can change the URL, this makes the cached `a` link unusable.

In order to prevent that, I need to reset the cache by defining the URL to any value, then I need to reset again to set the URL to `''` to support relative paths like `./image.png` to be correctly displayed.

URLs are not updated if you set the same value, that's why I need to reset the value twice.

JonasBa replied:
Makes perfect sense, I would add something similar as an inline comment

eoghanmurray commented:
@JonasBa Could you also test the following possibility:

```
function getHref(doc: Document) {
  // return a href without hash
  const url = new URL(doc.location.href);
  url.hash = ''
  return url.href;
}
```

Some more work would need to be done to support the new `customHref` case so maybe this PR is fine

YunFeng0817 commented on packages/rrweb-snapshot/src/snapshot.ts:
```diff
@@ -194,23 +194,45 @@ function getAbsoluteSrcsetString(doc: Document, attributeValue: string) {
-  const a: HTMLAnchorElement = doc.createElement('a');
-  a.href = attributeValue;
-  return a.href;
+
+  return getHref(doc, attributeValue);
```
> @H4ad Could you also test the following possibility:
> 
> ```
> function getHref(doc: Document) {
>   // return a href without hash
>   const url = new URL(doc.location.href);
>   url.hash = ''
>   return url.href;
> }
> ```
> 
> Some more work would need to be done to support the new `customHref` case so maybe this PR is fine

A slight change based on @eoghanmurray 's suggestion
```suggestion

  const url = new Url(attributeValue, doc.location);
  url.hash = '';
  return url.href;
```

eoghanmurray commented:
One way in which the `new URL` method throws an error, where the existing way doesn't is with `attributeValue = 'http://'`

We might want to use a fallback to the old method in that case.  Also not all browsers/platforms support the `URL` object.

eoghanmurray commented:
I've a version which should perform just as good without the need for caching in #1434 

@H4ad would you like to run benchmarks on that?

H4ad commented:
@eoghanmurray Sorry for taking so long to answer, I will run the benchmarks as soon I can and I will bring back the results

eoghanmurray commented:
(in case you miss the thread further up)

Could you cherry-pick https://github.com/eoghanmurray/rrweb/commit/615e164fe6bf3f48f5766f283e9e94d723cfc7fa

And also rebase or merge master

H4ad commented:
@eoghanmurray  Done!

eoghanmurray commented:
I've added another review comment as I'd like to be able to recreate the SPA scenario via a test which necessitates the cache bust.  I created a test case but was not able to recreate the scenario where the cache bust is needed.

H4ad commented:
@eoghanmurray did you submit the review?

eoghanmurray commented:
> I've added another review comment
This is what I was referring to:
https://github.com/rrweb-io/rrweb/pull/1387/files#diff-6434d77e679a3d1e4217dcad03ee6f02551e41dc0b077a06a45020e68e1b4ee3R222

Here is the test I was talking about, which could be cherry-picked in here, but basically it doesn't demonstrate any need for the `./clear-current` stuff, so maybe you can enhance the test case?
https://github.com/eoghanmurray/rrweb/commit/4d521651eb4c27e96a8c786860d44a386411da34

And here's some possible refactoring depending on the outcome of the above:
https://github.com/eoghanmurray/rrweb/commit/38e677a29ff23e8e91fc697aa9e01772140c3171

eoghanmurray commented:
I propose merging a version of my changes:
https://github.com/rrweb-io/rrweb/commit/38e677a29ff23e8e91fc697aa9e01772140c3171

but without the `This is needed for SPAs.` bit as we don't yet have any evidence that that is necessary.

H4ad commented:
You are right, I don't know if it was a bug on Chrome when I tested but the behavior that I described in the comment is not true anymore.

I have updated the code to reflect your changes and I also removed those comments/code about SPA.

Thanks for the patience, I was very lazy on this PR.

eoghanmurray commented on packages/rrweb-snapshot/src/snapshot.ts:
```diff
@@ -195,23 +195,34 @@ function getAbsoluteSrcsetString(doc: Document, attributeValue: string) {
-  a.href = '';
+function getHref(doc: Document, customHref?: string) {
+  let a = cachedDocument.get(doc);
+
```
```suggestion
```

eoghanmurray commented on packages/rrweb-snapshot/src/snapshot.ts:
```diff
@@ -195,23 +195,34 @@ function getAbsoluteSrcsetString(doc: Document, attributeValue: string) {
+    a = doc.createElement('a');
+    cachedDocument.set(doc, a);
+  }
+
```
```suggestion
```

eoghanmurray commented on packages/rrweb-snapshot/src/snapshot.ts:
```diff
@@ -195,23 +195,34 @@ function getAbsoluteSrcsetString(doc: Document, attributeValue: string) {
+  if (!customHref) {
+    customHref = '';
+  }
+
```
```suggestion
```

eoghanmurray commented on packages/rrweb-snapshot/src/snapshot.ts:
```diff
@@ -195,23 +195,34 @@ function getAbsoluteSrcsetString(doc: Document, attributeValue: string) {
+  }
+
+  a.setAttribute('href', customHref);
+
```
```suggestion
```

eoghanmurray commented on packages/rrweb-snapshot/src/snapshot.ts:
```diff
@@ -195,23 +195,30 @@ function getAbsoluteSrcsetString(doc: Document, attributeValue: string) {
+  if (!customHref) {
+    customHref = '';
+  }
+  a.setAttribute('href', customHref);
```
```suggestion
  // note: using `new URL` is slower https://jsbench.me/uqlud17rxo/1
  a.setAttribute('href', customHref);
```

eoghanmurray commented on packages/rrweb-snapshot/src/snapshot.ts:
```diff
@@ -195,23 +195,31 @@ function getAbsoluteSrcsetString(doc: Document, attributeValue: string) {
+  if (!customHref) {
+    customHref = '';
+  }
+  // note: using `new URL` is slower https://jsbench.me/uqlud17rxo/1
```
```suggestion
  // note: using `new URL` is slower. See #1434 or https://jsbench.me/uqlud17rxo/1
```

eoghanmurray commented:
Actually I'll just create a new PR for that test case

eoghanmurray commented:
@H4ad I found a bug where a `blob:` style url would be persisted on the cached anchor element and would incorrect return values on subsequent calls to getHref.  Fix in #1453 

But could you re-review this PR in view of those single page app bugs you previously encountered?

H4ad commented:
Sure, I will try leave a review as soon as possible.

eoghanmurray commented:
Just to note, using `URL.canParse` can make #1434 more competitve; only ~20% slower in jsbench:

```
function absoluteToDocCanParse(doc, attributeValue) {
// todo: do we need check if attributeValue is ''? 
    if (URL && URL.canParse(attributeValue)) {
	return new URL(attributeValue, doc.location.href).href;
    } else {
	const a = doc.createElement('a');
	a.setAttribute('href', attributeValue);
	return a.href;
    }
}
```
